pkg/reconciler/workload/virtualworkspaceurls/virtualworkspaceurls_reconcile.go

```diff
@@ -0,0 +1,58 @@
+/*
+Copyright 2022 The KCP Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package virtualworkspaceurls
+
+import (
+	"net/url"
+	"path"
+
+	"github.com/kcp-dev/logicalcluster"
+
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/klog/v2"
+
+	virtualworkspacesoptions "github.com/kcp-dev/kcp/cmd/virtual-workspaces/options"
+	"github.com/kcp-dev/kcp/pkg/apis/tenancy/v1alpha1"
+	workloadv1alpha1 "github.com/kcp-dev/kcp/pkg/apis/workload/v1alpha1"
+	synceroptions "github.com/kcp-dev/kcp/pkg/virtual/syncer/options"
+)
+
+func (c *Controller) reconcile(workloadCluster *workloadv1alpha1.WorkloadCluster, workspaceShards []*v1alpha1.ClusterWorkspaceShard) (*workloadv1alpha1.WorkloadCluster, error) {
+	workloadClusterCopy := workloadCluster.DeepCopy()
+
+	syncerVirtualWorkspaceURLs := sets.NewString()
+	for _, workspaceShard := range workspaceShards {
+		if workspaceShard.Spec.ExternalURL != "" {
+			syncerVirtualWorkspaceURL, err := url.Parse(workspaceShard.Spec.ExternalURL)
+			if err != nil {
+				klog.Errorf("failed to parse workspaceShard.Spec.BaseURL: %v", err)
+				return nil, err
+			}
+			syncerVirtualWorkspaceURL.Path = path.Join(
+				syncerVirtualWorkspaceURL.Path,
+				virtualworkspacesoptions.DefaultRootPathPrefix,
+				synceroptions.SyncerVirtualWorkspaceName,
+				logicalcluster.From(workloadClusterCopy).String(),
+				workloadClusterCopy.Name,
+			)
+			syncerVirtualWorkspaceURLs.Insert(syncerVirtualWorkspaceURL.String())
+		}
+	}
+	workloadClusterCopy.Status.SyncerVirtualWorkspaceURLs = syncerVirtualWorkspaceURLs.List()
```

Comment:
you have to sort the list I believe to make it the same on every reconcile, don't you?